cmd/subctl/upgrade.go

```diff
@@ -0,0 +1,271 @@
+//go:build !deploy
+
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subctl
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/google/go-github/v54/github"
+	"github.com/spf13/cobra"
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/subctl/internal/cli"
+	"github.com/submariner-io/subctl/internal/constants"
+	"github.com/submariner-io/subctl/internal/exit"
+	"github.com/submariner-io/subctl/internal/restconfig"
+	"github.com/submariner-io/subctl/pkg/cluster"
+	"github.com/submariner-io/subctl/pkg/deploy"
+	"github.com/submariner-io/subctl/pkg/version"
+	"github.com/submariner-io/submariner-operator/api/v1alpha1"
+)
+
+var (
+	to                        string
+	upgradeRestConfigProducer = restconfig.NewProducer()
+)
+
+// upgradeCmd represents the upgrade command.
+var upgradeCmd = &cobra.Command{
+	Use:   "upgrade",
+	Short: "Upgrades Submariner",
+	Run:   upgrade,
+}
+
+func init() {
+	upgradeCmd.Flags().StringVar(&to, "to", "", "the version of Submariner to upgrade to")
+	upgradeRestConfigProducer.SetupFlags(upgradeCmd.Flags())
+	rootCmd.AddCommand(upgradeCmd)
+}
+
+func upgrade(_ *cobra.Command, _ []string) {
+	status := cli.NewReporter()
+
+	// Step 1: upgrade subctl to match the requested version
+	command, err := upgradeSubctl(status)
+	exit.OnError(err)
+
+	if command != nil {
+		// Step 2a: subctl was upgraded, so run it instead of continuing
+		cmd := exec.Cmd{
+			Path:   *command,
+			Args:   os.Args[1:],
+			Stdin:  os.Stdin,
+			Stdout: os.Stdout,
+			Stderr: os.Stderr,
+		}
+		// exit.OnError outputs the version of subctl, which ends up being confusing here
+		if err := cmd.Run(); err != nil {
+			os.Exit(1)
+		}
+	} else {
+		// Step 2b: this subctl is already the requested version, run it
+		exit.OnError(upgradeRestConfigProducer.RunOnAllContexts(upgradeSubmariner, status))
+	}
+}
+
+// upgradeSubctl upgrades the local copy of subctl, if necessary.
+// Returns the path to the upgraded subctl if subctl was upgraded, nil if it wasn't.
+func upgradeSubctl(status reporter.Interface) (*string, error) {
```

Comment:
Minor thing but I think it's more idiomatic to return an empty string rather than nil, ie an empty string is equivalent to nil.

```suggestion
func upgradeSubctl(status reporter.Interface) (string, error) {
```